Refactor object internal

clementetb opened the pull request:
Moves all properties from `RealmObjectInternal` into the composite object `ObjectReference`. By doing this we can enforce the invariant that only an object with the composite object is a managed object, and thus all its properties must be non-null.

Tasks
- [x] Refactor base code.
- [x] Update compiler plugin to use the new composite object and adopt invariants.
- [x] Remove now unnecessary nullability asserts.
- [x] Rename properties, no need to hide them behind a dollar sign.
- [x] Homogenise object creation and linking.

This PR does not try to optimize the access to the the native pointer, neither tries to merge `RealmObjectHelper` in `ObjectReference`.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectInternal.kt:
```diff
@@ -43,130 +31,36 @@ import kotlin.reflect.KClass
     }
 }
+
+internal fun RealmObjectInternal.propertyInfoOrThrow(
```
Why do we need those? My thought was that we should guard all entry points with `isManaged` (or something like that) and then only work with `ObjectReference`s internally. Maybe by a pattern like:
```
realmObject.asObjectRefernce?.<method> {
   ...
} ?: throw Illegal ...
```
Maybe I need to check it out and play with it 🤔

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/dynamic/DynamicMutableRealmImpl.kt:
```diff
@@ -79,10 +82,10 @@ internal open class DynamicMutableRealmImpl(
             null
         } else if (!obj.isManaged()) {
```
If we rearranged this to use `obj.asObjectReference()?. { ... } ?: throw ` then we have immediate access to all the internals in the block and don't need to mess with `RealmObjectInternal` too!?

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/RealmObject.kt:
```diff
@@ -40,14 +41,22 @@ public interface RealmObject : Deleteable
  */
 public fun RealmObject.version(): VersionId {
-    return realmObjectInternal().version()
+    if (!isManaged()) {
```
Can't we just dispatch it immediately 
```
return asObjectReference?.let { version() } ?: UnmanagedVersion.version()
```
or even 
```
 return (asObjectRefence ?: UnmanagedVersion).version()
``` 
🙈

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/RealmObject.kt:
```diff
@@ -74,9 +83,9 @@ internal fun RealmObject.hasSameObjectKey(other: RealmObject?): Boolean {
 
     val thisKey =
-        RealmInterop.realm_object_get_key(this.realmObjectInternal().`$realm$ObjectPointer`!!)
+        RealmInterop.realm_object_get_key(this.asObjectReference()!!.objectPointer)
```
Again, if we reworked the above `isManaged` into `asObjectReference()?.<some-method>` then we wouldn't have to handle all the unmanaged state (`!!`, etc.) here.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/ObjectReference.kt:
```diff
@@ -0,0 +1,150 @@
+    Observable<ObjectReference<out RealmObject>, ObjectChange<out RealmObject>>,
+    Flowable<ObjectChange<out RealmObject>> {
+
+    public lateinit var owner: RealmReference
```
All these properties should just be initialized in the constructor now. We would never have an `ObjectReference` without actually having a managed object as it was the case for all these properties that was also present in the `RealmObject`.

clementetb replied:
I left like that because these properties are being directly initialized on the macos `InstrumentedTests`. I have added a secondary constructor to support this test case.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt:
```diff
@@ -32,17 +31,21 @@ internal fun <T : RealmObject> RealmObjectInternal.manage(
-        RealmInterop.realm_get_class(`$realm$Owner`!!.dbPointer, RealmInterop.realm_object_get_table(objectPointer)).name
-    } else {
-        realmObjectCompanionOrThrow(type).`$realm$className`
+    this.`$realm$objectReference` = ObjectReference(type).apply {
```
Can't we move all this to the constructor of `ObjectReference`. All the properties can now be `val`s that are initialized in the constructor, right? Then we can enforce the invariants that we never operate on `ObjectReference`s without them being fully initialized!?

rorbech reviewed (commented):
Seems to be on the right path with a lot of simplications, but can't we just do the 
```
this.`$realm$objectReference`?.let { it.getValue("propertyName") } ?: backingField
```
indirection directly in the IR, so we always just operate directly on the `ObjectReference`. There is no need to go into internal code for the unmanaged objects and it would simplify the internal code drastically if we should handle the unmanaged cases.

cmelchior commented on packages/cinterop/src/jvm/kotlin/io/realm/internal/interop/RealmInterop.kt:
```diff
@@ -501,8 +501,7 @@ actual object RealmInterop {
                 is RealmObjectInterop -> {
-                    val nativePointer = (value as RealmObjectInterop).`$realm$ObjectPointer`
-                        ?: error("Cannot add unmanaged object")
+                    val nativePointer = (value as RealmObjectInterop).objectPointer
```
Isn't the cast automatic inside the `is RealmObjectInterop` ?

clementetb replied:
Yes, it is done automatically by the when. I didn't noticed that it could be removed.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/RealmObject.kt:
```diff
@@ -88,7 +97,7 @@ internal fun RealmObject.hasSameObjectKey(other: RealmObject?): Boolean {
     return if (isManaged()) {
         val internalObject = this as RealmObjectInternal
-        val ptr = internalObject.`$realm$ObjectPointer`
+        val ptr = internalObject.`$realm$objectReference`!!.objectPointer
```
I don't have the full overview, but we have these nullchecks all over the place, which looks a bit glaring to me and indicates there is some state or type we haven't nailed just yet. It would be nice if we could somehow get rid of this.

clementetb replied:
Yes, claus also suggested it. Now I do the check implicitly when extracting the `ObjectReference`.

cmelchior commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/ObjectReference.kt:
```diff
@@ -0,0 +1,150 @@
+import kotlinx.coroutines.flow.Flow
+import kotlin.reflect.KClass
+
+public class ObjectReference<T : RealmObject>(internal val type: KClass<T>) :
```
Missing class doc for what this class does.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/query/SingleQuery.kt:
```diff
@@ -100,6 +103,6 @@ internal class SingleQuery<E : RealmObject> constructor(
         // TODO C-API doesn't implement realm_query_delete_all so just fetch the result and delete
         //  that
-        find()?.asInternalDeleteable()?.delete()
+        find()?.asObjectReference()?.asInternalDeleteable()?.delete()
```
`ObjectReference` already implements the `InternalDeletable`, we can call it directly.

clementetb commented:
Since the last reviews:

- Renamed `ObjectReference` to `RealmObjectReference`
- Renamed `asObjectReference` to `getObjectReference`
- IR modified accessors pass the `RealmObjectReference` instead of a `RealmObjectInternal`
- Introduced `RealmObjectReference.runIfManaged` same as: `getObjectReference()?.run{}`

rorbech reviewed (changes requested):
Really nice. 

But ... I am a concerned with the definition of `runIfManaged` returning `R?` which can either be null due to `getObjectReference()` returning `null` or the result of `block`. I think this can lead to unexpected scenarios. I think we need to differentiate it into different methods where it is explicit whether the result is the return value from the block, etc.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/InternalDeleteable.kt:
```diff
@@ -24,10 +24,11 @@ public interface InternalDeleteable : Deleteable {
 
 internal fun Deleteable.asInternalDeleteable(): InternalDeleteable {
-    if (this !is InternalDeleteable) {
+    val deleteable = if (this is RealmObjectInternal) { this.getObjectReference() } else { this }
```
It is not possible to rearrange it. There are no `RealmObject.asInternalDeleteable()` as `Deleteable.asInternalDeleteable()` is always invoked with a parameter of type `Deleteable` see [InternalMutableRealm.kt#L64](https://github.dev/realm/realm-kotlin/blob/aaf683554aad25a2a2c8185c5c195db88407fe2a/packages/library-base/src/commonMain/kotlin/io/realm/internal/InternalMutableRealm.kt#L64)

We always need to differentiate between these two, or make `RealmObjectInternal` to implement `Deleteable` and add a fake override in the compiler plugin.

clementetb replied:
I will add a comment to explain this "complicated" code.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/InternalMutableRealm.kt:
```diff
@@ -30,18 +29,20 @@ internal interface InternalMutableRealm : MutableRealm {
+                    // up to date, just return input
+                    obj
+                } else {
+                    return obj.thaw(realmReference) as T?
```
Good call, then we can remove the extension method `thaw`

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmListInternal.kt:
```diff
@@ -80,7 +80,12 @@ internal class ManagedRealmList<E>(
                 index.toLong(),
-                copyToRealm(metadata.mediator, metadata.realm, element)
+                copyToRealm(metadata.mediator, metadata.realm, element).let { value ->
+                    when (value) {
```
Maybe this can just be handled in the `ElementConverter` for RealmObjects 🤔. If not then maybe it is better left to be reworked as part of #728

clementetb replied:
At least we can move the let part into a helper function.

clementetb replied:
Finally I am leaving a `FIXME` pointing to https://github.com/realm/realm-kotlin/issues/728

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:
```diff
@@ -52,87 +51,75 @@ internal object RealmObjectHelper {
-        // TODO Error could be eliminated if we only reached here on a ManagedRealmObject (or something like that)
-        val o = obj.`$realm$ObjectPointer`!!
-            ?: sdkError("Cannot retrieve property value in a realm for an unmanaged objects")
+        val o = obj.objectPointer
```
Actually there are more variables that can be inlined and single-line functions.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/dynamic/DynamicMutableRealmImpl.kt:
```diff
@@ -77,12 +78,16 @@ internal open class DynamicMutableRealmImpl(
+                if (owner == realmReference) {
+                    obj as DynamicMutableRealmObject?
+                } else {
+                    return obj.getObjectReference()!!
```
Oh I missed changing it.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt:
```diff
@@ -101,3 +130,38 @@ internal inline fun <reified T : Any> KClass<T>.realmObjectCompanionOrNull(): Re
+        val that = this
+        other.runIfManaged {
+            val thisKey =
+                RealmInterop.realm_object_get_key(this.objectPointer)
+            val otherKey =
+                RealmInterop.realm_object_get_key(that.objectPointer)
+
```
It is not possible to use the `this@` as there is a name clash.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectInternal.kt:
```diff
@@ -43,130 +27,6 @@ import kotlin.reflect.KClass
  */
 // TODO Public due to being a transative dependency of Mediator
```
That dependency goes down to the current sync split. They are needed to be public so they are accessible from the sync module.

We could instead of having base a dependency on Sync, to link the base sources in the sync project. It then would have access to all the internals.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt:
```diff
@@ -61,20 +64,18 @@ internal fun <T : RealmObject> RealmObjectInternal.link(
-    return this.freeze(frozenRealm) as T
-}
+@Suppress("UNCHECKED_CAST")
+internal fun <T : RealmObject> RealmObject.freeze(frozenRealm: RealmReference): T =
```
It is only used once, I am removing it and inlining the logic.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/dynamic/DynamicMutableRealmObjectImpl.kt:
```diff
@@ -35,11 +34,11 @@ internal class DynamicMutableRealmObjectImpl : DynamicMutableRealmObject, Dynami
         when (value) {
             is RealmObject -> RealmObjectHelper.setObject(
-                this,
+                this.`$realm$objectReference`!!,
```
The dynamic object is already managed, I didn't found any way to force the object reference to be non null. Leaving a comment stating the invariant.

clementetb commented on packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt:
```diff
@@ -376,139 +380,161 @@ class AccessorModifierIrGeneration(private val pluginContext: IrPluginContext) {
-        // CALL 'public open fun <get-isManaged> (): kotlin.Boolean declared in io.realm.example.Sample' type=kotlin.Boolean origin=GET_PROPERTY
-        return irCall(
-            isManagedProperty.getter!!,
+    private fun IrStatementsBuilder<*>.irObjectReferenceBlock(
```
Much better 💎

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt:
```diff
@@ -101,3 +130,38 @@ internal inline fun <reified T : Any> KClass<T>.realmObjectCompanionOrNull(): Re
+/**
+ * If the Realm object is managed returns its Realm Object reference, otherwise returns null.
+ */
+internal fun RealmObject.getObjectReference(): RealmObjectReference<out RealmObject>? {
```
I believe that you refer to `getOrNull`, it is part of the collection and exists to provide a way around if the index being accessed is out of bounds, it catches the error and returns null.

I don't see the need to state that it is nullable, as the return type already tells about it 🤷‍♂️

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt:
```diff
@@ -101,3 +130,38 @@ internal inline fun <reified T : Any> KClass<T>.realmObjectCompanionOrNull(): Re
+                RealmInterop.realm_object_get_key(that.objectPointer)
+
+            thisKey == otherKey
+        }
```
Then it would return null, and thus would hit the throw.

clementetb commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectHelper.kt:
```diff
@@ -225,29 +212,33 @@ internal object RealmObjectHelper {
-            } else value
-        setValueByKey(obj, obj.propertyInfoOrThrow(propertyName).key, newValue)
+        val newValue = value?.runIfManaged {
+            if (obj.owner != owner) {
```
Yes there is a way 🎯

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectInternal.kt:
```diff
@@ -43,130 +31,36 @@ import kotlin.reflect.KClass
-        return this.`$realm$Owner`!!.owner.registerObserver(this)
-    }
+public interface RealmObjectInternal : RealmObject {
+    public var `$realm$objectReference`: ObjectReference<out RealmObject>? // RealmObjectReference ?
```
We can change that as part of #763

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/dynamic/DynamicRealmObjectImpl.kt:
```diff
@@ -17,44 +17,39 @@
-    override var `$realm$metadata`: ClassMetadata? = null
+        get() = this.getObjectReference()!!.className
+
+    override var `$realm$objectReference`: RealmObjectReference<out RealmObject>? = null
```
See https://github.com/realm/realm-kotlin/pull/736#discussion_r835296369

cmelchior commented on packages/plugin-compiler/src/main/kotlin/io/realm/compiler/AccessorModifierIrGeneration.kt:
````diff
@@ -376,17 +380,37 @@ class AccessorModifierIrGeneration(private val pluginContext: IrPluginContext) {
+             * Transform the getter to whether access the managed object or the backing field
+             * ```
+             * get() {
+             *      return this.`$realm$objectReference`?.let { it.getValue("propertyName") } ?: backingField
````
While this is a really nice one-liner, we should probably at some point check exactly what is generated in JVM for this. Especially since this is our hottest path.

rorbech replied:
I played around with various approaches and already inspected the IR dump and it seems quite reasonable to me: https://github.com/realm/realm-kotlin/blob/ct/reafactor-object-internal/packages/plugin-compiler/src/test/resources/sample/expected/00_ValidateIrBeforeLowering.ir#L117..L135

cmelchior reviewed (approved):
I'm probably missing some nuances, but generally this looks like some really good optimizations.

rorbech commented on packages/library-base/src/commonMain/kotlin/io/realm/internal/RealmObjectUtil.kt:
```diff
@@ -32,17 +31,21 @@ internal fun <T : RealmObject> RealmObjectInternal.manage(
+                RealmInterop.realm_object_get_table(objectPointer)
+            ).name
+        } else {
+            realmObjectCompanionOrThrow(type).`$realm$className`
```
Same
